### PR TITLE
feat: Create new endpoint/entity for ExpressionDimensionItem [2.40-DHIS2-13988]

### PR DESCRIPTION
**Pull request by d-bernat**
please see: https://dhis2.atlassian.net/jira/software/c/projects/DHIS2/boards/104?modal=detail&selectedIssue=DHIS2-13988&quickFilter=497

### CODE REVIEW

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/ExpressionDimensionItemController.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+/**
+ * @author Dusan Bernat
+ */
+
+/**
+ *
+ */
```
Comment (gnespolino) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/ExpressionDimensionItemController.java:61]: please remove this empty comment

Comment (gnespolino) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/ExpressionDimensionItemController.java:61]: please remove this empty comment

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/ExpressionDimensionItemController.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+     * @return
+     */
+    @ResponseBody
+    @PostMapping( produces = APPLICATION_JSON_VALUE )
+    public ResponseEntity<ExpressionDimensionItem> createExpressionDimensionItem(
+        @RequestBody ExpressionDimensionItem expressionDimensionItem )
+    {
+        try
+        {
+            expressionService.addExpressionDimensionItem( expressionDimensionItem );
```
Comment (gnespolino) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/ExpressionDimensionItemController.java:141]: is it safe to add an expressionItem without checking if another with the same id already exists?

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/ExpressionDimensionItemController.java [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+     * @return
+     */
+    @ResponseBody
+    @PutMapping( produces = APPLICATION_JSON_VALUE )
+    public ResponseEntity<ExpressionDimensionItem> updateExpressionDimensionItem(
+        @RequestBody ExpressionDimensionItem expressionDimensionItem )
+    {
+        try
+        {
+            expressionService.updateExpressionDimensionItem( expressionDimensionItem );
```
Comment (gnespolino) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/ExpressionDimensionItemController.java:163]: is it safe to update an expressionDimensionItem if the id doesn't exist?

**Review by gnespolino — APPROVED** [review]

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/expressiondimensionitem/ExpressionDimensionItem.java [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+ * i) [1.2] where 1 refers to the data element identifier and 2 refers to the
+ * category option combo identifier.
+ * <p/>
+ * ii) [1] where 1 refers to the data element identifier, in this case the
+ * formula represents the total value for all category option combos for that
+ * data element.
+ *
+ */
+@JacksonXmlRootElement( localName = "expressionDimensionItem", namespace = DxfNamespaces.DXF_2_0 )
+public class ExpressionDimensionItem
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/expressiondimensionitem/ExpressionDimensionItem.java:59]: I think we should extend `BaseIdentifiableObject`, so you get all required default/standard attributes for free.
These attributes should be added as columns to your table, respecting the same column types and max chars from existing tables.
You should get for free equals/hashcode, translation, etc.

**Review by maikelarabori — CHANGES_REQUESTED** [review]
Left some comments that will probably require changes @d-bernat.

Thanks!

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java [hunk 5]
```diff
@@ -278,8 +278,16 @@ protected void addTransientOrganisationUnit( OrganisationUnit organisationUnit )
      */
     public List<DimensionalItemObject> getDataDimensionNameableObjects()
     {
-        return dataDimensionItems.stream().map( DataDimensionItem::getDimensionalItemObject )
```
Comment (gnespolino) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java:281]: it was enough to add `.filter(Objects::nonNull)` in the stream chain.

**Review by maikelarabori — CHANGES_REQUESTED** [review]
Hey @d-bernat, looks good in general :+1: 
I left a few comments that would be nice to be reviewed before merging it.

Thanks!

**Review by maikelarabori — APPROVED** [review]

**Pull request merged**